README.md


[![Greenkeeper badge](https://badges.greenkeeper.io/carpages/gemini-carousel.svg)](https://greenkeeper.io/)

CHANGELOG.md
# CHANGELOG

## v1.1.2

  - Deprecate PlaceTim urls for testing
  - Bump Gemini dependency versions
    - gemini-touch - Bump HammerJS -> 2.0

## v1.1.1

  - Update changelog

## v1.1.0

  - Update Handlebars to v4
  - Fix loading of QUnit ( via PhantomJS ) in unit tests

## v1.0.7

  - Add a public update method

## v1.0.6

  - Fix math for looping to last slide from first slide

## v1.0.5

  - Update CommonJS require to proper module name

## v1.0.4

  - Fix linting grunt task
  - Fix linting errors

## v1.0.3

  - Remove jshint in favour of eslint with custom ruleset
  - Bump travis-ci node version
  - Fix CommonJS require statement to require the correct gemini-loader module

## v1.0.2

  - Fix looping to work both ways

## v1.0.1

  - Fix Handlebar templates
  - Update package.json with proper dependencies

## v1.0.0

  - Requires Handlebars v3.0.3 now
  - Extension is put on .carousel instead of .carousel__list now
    - Example: ``.carousel--2``


test/qunit.config.js
// RequireJS config for Tests
requirejs.config({
  baseUrl: '../',
  paths: {
    'qunit': 'bower_components/qunit/qunit/qunit',
    'underscore': 'bower_components/underscore/underscore',
    'jquery': 'bower_components/jquery/dist/jquery',
    'handlebars': 'bower_components/handlebars/handlebars.runtime',
    'jquery.boiler': 'bower_components/jquery-boiler/jquery.boiler',
    'gemini.support': 'bower_components/gemini-support/gemini.support',
    'gemini': 'bower_components/gemini-loader/gemini',
    'gemini.fold': 'bower_components/gemini-fold/gemini.fold',
    'gemini.respond': 'bower_components/gemini-respond/gemini.respond',
    'gemini.touch': 'bower_components/gemini-touch/gemini.touch',
    'fastclick': 'bower_components/fastclick/lib/fastclick',
    'hammerjs': 'bower_components/hammerjs/hammer',
    'jquery.hammer': 'bower_components/jquery-hammerjs/jquery.hammer',
    'gemini.carousel.templates': 'templates'
  }
});


test/gemini-carousel.edge.js
/* eslint key-spacing: ["error", { "align": "colon" }] */
require.config({
  baseUrl: '../',
  paths  : {
    'jquery'                   : 'bower_components/jquery/dist/jquery',
    'jquery.boiler'            : 'bower_components/jquery-boiler/jquery.boiler',
    'handlebars'               : 'bower_components/handlebars/handlebars.runtime',
    'underscore'               : 'bower_components/underscore/underscore',
    'gemini'                   : 'bower_components/gemini-loader/gemini',
    'gemini.fold'              : 'bower_components/gemini-fold/gemini.fold',
    'gemini.respond'           : 'bower_components/gemini-respond/gemini.respond',
    'gemini.support'           : 'bower_components/gemini-support/gemini.support',
    'gemini.carousel.templates': 'templates'
  }
});

require([ 'gemini', 'gemini.carousel' ], function( G ) {
  $( '.carousel' ).carousel();
});


Gruntfile.js
'use strict';

var compassImporter = require( 'compass-importer' );

module.exports = function( grunt ) {
  // Load all grunt tasks
  require( 'load-grunt-tasks' )( grunt );

  // Show elapsed time at the end
  require( 'time-grunt' )( grunt );

  // Project configuration.
  grunt.initConfig({
    // Metadata.
    pkg: grunt.file.readJSON( 'package.json' ),
    banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' +
      '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
      '<%= pkg.homepage ? "* " + pkg.homepage + "\\n" : "" %>' +
      '* Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>;' +
      ' Licensed MIT */\n',
    // Task configuration.
    handlebars: {
      compile: {
        options: {
          umd: true,
          namespace: 'Templates.Default.Carousel',
          processName: function( filename ) {
            return filename.split( '/' ).pop().replace( '.hbs', '' );
          }
        },
        files: [{
          src: 'templates/*.hbs',
          dest: 'templates.js'
        }]
      }
    },
    qunit: {
      all: {
        options: {
          inject: [
            './test/qunit.config.js',
            './node_modules/grunt-contrib-qunit/phantomjs/bridge.js'
          ],
          timeout: 10000,
          urls: [ 'http://localhost:9000/test/<%= pkg.name %>.test.html' ],
          page: {
            viewportSize: {
              width: 10000,
              height: 10000
            }
          }
        }
      }
    },
    eslint: {
      options: {
        configFile: '.eslintrc'
      },
      target: [ 'gemini.carousel.js' ]
    },
    connect: {
      server: {
        options: {
          hostname: '*',
          base: '.',
          port: 9000
        }
      }
    },
    sass: {
      options: {
        importer: compassImporter,
        includePaths: [ 'bower_components' ]
      },
      dist: {
        options: {
          outputStyle: 'expanded'
        },
        files: {
          'test/gemini-carousel.test.css': 'test/gemini-carousel.test.scss'
        }
      }
    },
    'saucelabs-qunit': {
      all: {
        options: {
          urls: [ 'http://localhost:9000/test/<%= pkg.name %>.test.html' ],
          build: process.env.TRAVIS_JOB_ID,
          browsers: [
            // iOS
            {
              browserName: 'iphone',
              platform: 'OS X 10.9',
              version: '7.1'
            },
            {
              browserName: 'ipad',
              platform: 'OS X 10.9',
              version: '7.1'
            },
            // Android
            {
              browserName: 'android',
              platform: 'Linux',
              version: '4.3'
            },
            // OS X
            {
              browserName: 'safari',
              platform: 'OS X 10.9',
              version: '7'
            },
            {
              browserName: 'safari',
              platform: 'OS X 10.8',
              version: '6'
            },
            {
              browserName: 'firefox',
              platform: 'OS X 10.9',
              version: '28'
            },
            // Windows
            {
              browserName: 'internet explorer',
              platform: 'Windows 8.1',
              version: '11'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 8',
              version: '10'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '11'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '10'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '9'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '8'
            },
            {
              browserName: 'firefox',
              platform: 'Windows 7',
              version: '29'
            },
            {
              browserName: 'chrome',
              platform: 'Windows 7',
              version: '34'
            },
            // Linux
            {
              browserName: 'firefox',
              platform: 'Linux',
              version: '29'
            }
          ]
        }
      }
    }
  });

  // Default task.
  grunt.registerTask( 'default', [ 'sass', 'eslint', 'test' ]);
  grunt.registerTask( 'test', [ 'connect', 'qunit' ]);
  grunt.registerTask( 'ci', [ 'default' ]);
};


templates.js
(function(factory) {

if (typeof define === 'function' && define.amd) {

define(['handlebars'], factory);

} else if (typeof exports === 'object') {

module.exports = factory(require('handlebars'));

} else {

factory(Handlebars);

}

}(function(Handlebars) {

this["Templates"] = this["Templates"] || {};
this["Templates"]["Default"] = this["Templates"]["Default"] || {};
this["Templates"]["Default"]["Carousel"] = this["Templates"]["Default"]["Carousel"] || {};

this["Templates"]["Default"]["Carousel"]["nav"] = Handlebars.template({"compiler":[7,">= 4.0.0"],"main":function(container,depth0,helpers,partials,data) {
    var helper;

  return "<ol class=\"carousel__nav pagination pagination--carousel\">\n\n 	<li class=\"pagination__prev\"><a href=\"#\" data-goto=\"--\">Previous</a></li>\n\n	<li class=\"pagination__item\">\n		<span class=\"carousel__current-page-count\">1</span>\n		/\n		<span class=\"carousel__page-count\">"
    + container.escapeExpression(((helper = (helper = helpers.pageCount || (depth0 != null ? depth0.pageCount : depth0)) != null ? helper : helpers.helperMissing),(typeof helper === "function" ? helper.call(depth0 != null ? depth0 : {},{"name":"pageCount","hash":{},"data":data}) : helper)))
    + "</span>\n	</li>\n\n	<li class=\"pagination__next\"><a href=\"#\" data-goto=\"++\">Next</a></li>\n\n</ol>\n";
},"useData":true});

return this["Templates"]["Default"]["Carousel"];

}));

jquery.carousel.js
/**
 * @fileoverview

A jQuery plugin to build dynamic carousel's. The plugin works
mostly with CSS, meaning the markup is quite manipulatable.

### Notes
- Requires an include to ``carousel.scss`` in your Gemini build

### Features
- You can set the number of items per page in the markup's list. This is set
using the CSS extension ``carousel__list--{number}``.
- You can add custom navigation buttons within the carousel using the
``data-goto`` attribute. The value set will take you to the corresponding page.
You can see this in the example

 *
 * @namespace jquery.carousel
 * @copyright Carpages.ca 2014
 * @author Matt Rose <matt@mattrose.ca>

 * @requires jquery
 * @requires jquery.boiler
 * @requires jquery.fold
 * @requires jquery.respond
 * @requires gemini
 * @requires gemini.touch
 * @requires gemini.support

 * @prop {boolean} pagination {@link jquery.carousel#pagination}
 * @prop {boolean} loop {@link jquery.carousel#loop}
 * @prop {string} container {@link jquery.carousel#container}
 * @prop {integer} indexList {@link jquery.carousel#indexList}
 * @prop {integer} scrollEventDelay {@link jquery.carousel#scrollEventDelay}
 * @prop {object} templates {@link jquery.carousel#templates}

 * @example
  <html>
    <div id="js-carousel-example" class="carousel">
      <ul class="carousel__list carousel__list--3">
        <li>Content for item 1</li>
        <li>Content for item 2</li>
        <li>Content for item 3</li>
        <li>Content for item 4</li>
      </ul>
      <!-- Can be added automatically using the pagination setting -->
      <button class="button" data-goto="--">Previous Page</button>
      <button class="button" data-goto="2">Page 2</button>
      <button class="button" data-goto="++">Next Page</button>
    </div>
  </html>

 * @example
  $('#js-carousel-example').carousel();
 */

define([
    'jquery-loader',
    'gemini-loader',
    'underscore',
    'jquery.carousel.templates',
    'jquery.boiler',
    'jquery.fold',
    'jquery.respond',
    'gemini.support',
    'gemini.touch'
  ], function($, GEM, _, T){

  $.boiler('carousel', {

    defaults: {
      /**
       * Whether to append pagination to the the carousel.
       * @name jquery.carousel#pagination
       * @type Boolean
       * @default false
       */
      pagination: false,

      /**
       * Whether you want the carousel to loop.
       * @name jquery.carousel#loop
       * @type Boolean
       * @default false
       */
      loop: false,

      /**
       * Selector for the carousel's container. If false, carousel is $el.
       * @name jquery.carousel#container
       * @type String
       * @default false
       */
      container: false,

      /**
       * Specify which iteration of the .carousel__list you'd like to index for
       * the carousel.
       * @name jquery.carousel#indexList
       * @type Interger
       * @default 0
       */
      indexList: 0,

      /**
       * The delay until the scroll event is triggered on the carousel after
       * click.
       * @name jquery.carousel#scrollEventDelay
       * @type Integer
       * @default 0
       */
      scrollEventDelay: 0,

      /**
       * Precompiled Handlebar templates to replace default. Expecting 'nav' for
       * default navigation.
       * @name jquery.carousel#templates
       * @type Object
       * @default {}
       */
      templates: {}
    },

    // Event listeners
    events: {
      'click [data-goto]': '_handleClick'
    },

    // Initiate the plugin
    init: function(){
      var P = this;

      //Extend the templates
      P.T = $.extend(T, P.settings.templates);

      // Init variables
      P.currentItem = P.currentPage = 1;

      // Cache jQuery objects
      P.$carousel = !!P.settings.container ?
                    P.$el.find(P.settings.container) :
                    P.$el;
      P.$carouselLists = P.$el.find('.carousel__list');
      P.carouselList = P.$carouselLists[P.settings.indexList];
      P.$carouselList = $(P.carouselList);
      P.$currentPage = P.$el.find('[data-goto="1"]');

      // Update
      P._update();

      // Update on resize
      $.respond.bind('resize', function(e, scrn){
        P._update();
        P.gotoPage(P.currentPage);
      });

      // Touch Support
      if(GEM.support.touch) {
        P._initTouch();
      }

    },

    /**
     * Update the carousel's cached values and render and templates
     * @name jquery.carousel#_update
     * @private
     * @function
     */
    _update: function(){
      var P = this;

      //Update Cache
      P.pageWidth = P.$carouselList.width();
      P.itemWidth = P.$carouselList.children('li:first-child').width();
      P.itemsPerPage = Math.round(P.pageWidth/P.itemWidth);

      P.itemCount = P.$carouselList.children('li').length;
      P.pageCount = Math.ceil( P.itemCount / P.itemsPerPage );

      // Template the pagination
      if(P.settings.pagination){
        P._paginate();
      }
    },

    /**
     * Render the pagination
     * @name jquery.carousel#_paginate
     * @private
     * @function
     */
    _paginate: function(){
      var P = this;

      //Template
      P.pagination = P.T.nav({
        'pageCount': P.pageCount
      });

      //Render
      if(!P.$pagination){
        //First load
        P.$pagination = $('<span/>').append(P.pagination);
        P.$el.append(P.$pagination);
      } else {
        //Replace exisiting
        P.$pagination.html(P.pagination);
      }

      //Cache
      P.$currentPageCount = P.$pagination.find('.carousel__current-page-count');
    },

    /**
     * Callback when a user clicks the carousel
     * @name jquery.carousel#_handleClick
     * @private
     * @function
     * @param {event#object} e Click event object
     * @param {element} target The targeted element
     */
    _handleClick: function(e, target){
      var P = this;
      var goTo = $(target).data('goto');

      if(!goTo) return;

      e.preventDefault();
      //Next
      if(goTo == "next" || goTo == "++")
        P.next();
      //Previous
      else if(goTo == "prev" || goTo == "previous" || goTo == "--")
        P.previous();
      //Go to page
      else P.gotoPage(goTo);
    },

    /**
     * Go to the next page on the carousel
     * @name jquery.carousel#next
     * @function
     */
    next: function(){
      var P = this;
      P.gotoPage(P.currentPage + 1);
    },

    /**
     * Go to the previous page on the carousel
     * @name jquery.carousel#previous
     * @function
     */
    previous: function(){
      var P = this;
      P.gotoPage(P.currentPage - 1);
    },

    /**
     * Go to a specific item in the carousel
     * @name jquery.carousel#_gotoItem
     * @private
     * @function
     */
    _gotoItem: function(item, animate){
      if(animate===undefined) animate = true;

      var P = this;
      if(item > P.itemCount) return;

      //Calculate the x offset in pixels
      var $item = P.$carouselList.children('li:nth-child('+item+')'),
          xOffset = $item.offset().left -
                    P.$carouselList.offset().left +
                    P.carouselList.scrollLeft;

      //Whether there are more items to scroll to
      var isMoreItems = $.rightoffold(
        P.$carouselList.find('li').last(),{
          container:P.$carouselList,
          threshold:-P.itemWidth-20
        }
      );

      //Make sure there's something to scroll to
      if ((item > P.currentItem && !isMoreItems) || xOffset < 0) return;

      //Change the item
      P.currentItem = item;
      if(animate){
        P.$carouselLists.animate({
          scrollLeft:xOffset
        }, 500);
      }else{
        P.$carouselLists.scrollLeft(xOffset);
      }

      setTimeout(
        _.bind(function(){
          $(this).trigger("scroll");
        }, P.$carousel),
      P.settings.scrollEventDelay);
    },

    /**
     * Go to a specific page in the carousel
     * @name jquery.carousel#gotoPage
     * @function
     * @param {Integer} page The desired page
     */
    gotoPage: function(page, animate){
      if(animate===undefined) animate = true;

      var P = this;
      if(page > P.pageCount) {
        if (P.settings.loop) P.gotoPage(1);
        return;
      }

      var item = P.itemsPerPage * (page - 1) + 1;

      P._gotoItem(item, animate);

      //Change active page
      P.currentPage = page;
      P.$currentPage.removeClass('is-active');
      P.$currentPage = P.$el.find('[data-goto="'+P.currentPage+'"]');
      P.$currentPage.addClass('is-active');
      if(!!P.$currentPageCount) P.$currentPageCount.html(page);
    },

    /**
     * Initiate the touch support for carousel's
     * @name jquery.carousel#_initTouch
     * @private
     * @function
     */
    _initTouch: function(){
      var P = this;

      require(['gemini.touch'], function(){

        //Add touch events
        P.$carouselList.hammer({
          drag_block_horizontal: true,
          drag_lock_to_axis: true,
          drag_lock_min_distance: 20,
          hold: false,
          tap: false
        }).on('release dragleft dragright swipeleft swiperight', function(ev){

          switch(ev.type) {
            case 'dragright':
            case 'dragleft':
              // stick to the finger
              var pageOffset = (P.currentPage-1)*P.pageWidth;
              var dragOffset = -ev.gesture.deltaX;

              // slow down at the first and last pane
              if(
                (P.currentPage == 1 && ev.gesture.direction == "right") ||
                (P.currentPage == P.pageCount && ev.gesture.direction == "left")
              ) {
                dragOffset *= 0.4;
              }

              P.$carouselLists.scrollLeft(pageOffset + dragOffset);
              break;

            case 'swipeleft':
              P.next();
              ev.gesture.stopDetect();
              break;

            case 'swiperight':
              P.previous();
              ev.gesture.stopDetect();
              break;

            case 'release':
              // more then 50% moved, navigate
              if(Math.abs(ev.gesture.deltaX) > P.pageWidth/2) {
                if(ev.gesture.direction == 'right') {
                  P.previous();
                } else {
                  P.next();
                }
              } else {
                P.gotoPage(P.currentPage);
              }
              break;
            }
        });
      });
    }

  });

  // Return the jquery object
  // This way you don't need to require both jquery and the plugin
  return $;

});


gemini.carousel.js
/**
 * @fileoverview

A Gemini plugin to build dynamic carousel's. The plugin works
mostly with CSS, meaning the markup is quite manipulatable.

### Notes
- Requires an include to ``carousel.scss`` in your Gemini build

### Features
- You can set the number of items per page in the markup's list. This is set
using the CSS extension ``carousel__list--{number}``.
- You can add custom navigation buttons within the carousel using the
``data-goto`` attribute. The value set will take you to the corresponding page.
You can see this in the example

 *
 * @namespace gemini.carousel
 * @copyright Carpages.ca 2014
 * @author Matt Rose <matt@mattrose.ca>

 * @requires gemini
 * @requires gemini.fold
 * @requires gemini.respond
 * @requires gemini.touch
 * @requires gemini.support

 * @prop {boolean} pagination {@link gemini.carousel#pagination}
 * @prop {boolean} loop {@link gemini.carousel#loop}
 * @prop {boolean} animate {@link gemini.carousel#animate}
 * @prop {string} container {@link gemini.carousel#container}
 * @prop {integer} indexList {@link gemini.carousel#indexList}
 * @prop {integer} scrollSpeed {@link gemini.carousel#scrollSpeed}
 * @prop {boolean} thumbs {@link gemini.carousel#thumbs}
 * @prop {function} onChange {@link gemini.carousel#onChange}
 * @prop {object} templates {@link gemini.carousel#templates}

 * @example
  <html>
    <div id="js-carousel-example" class="carousel">
      <ul class="carousel__list carousel__list--3">
        <li>Content for item 1</li>
        <li>Content for item 2</li>
        <li>Content for item 3</li>
        <li>Content for item 4</li>
      </ul>
      <!-- Can be added automatically using the pagination setting -->
      <button class="button" data-goto="--">Previous Page</button>
      <button class="button" data-goto="2">Page 2</button>
      <button class="button" data-goto="++">Next Page</button>
    </div>
  </html>

 * @example
  G('#js-carousel-example').carousel();
 */

define([
    'gemini',
    'gemini.carousel.templates',
    'gemini.fold',
    'gemini.respond',
    'gemini.support'
  ], function($, T){

  var _ = $._;

  $.boiler('carousel', {

    defaults: {
      /**
       * Whether to append pagination to the the carousel.
       * @name gemini.carousel#pagination
       * @type Boolean
       * @default false
       */
      pagination: false,

      /**
       * Whether to do a scroll animation when clicked
       * @name gemini.carousel#animate
       * @type Boolean
       * @default false
       */
      animate: true,

      /**
       * Whether you want the carousel to loop.
       * @name gemini.carousel#loop
       * @type Boolean
       * @default false
       */
      loop: false,

      /**
       * Selector for the carousel's container. If false, carousel is $el.
       * @name gemini.carousel#container
       * @type String
       * @default false
       */
      container: false,

      /**
       * Specify which iteration of the .carousel__list you'd like to index for
       * the carousel.
       * @name gemini.carousel#indexList
       * @type Interger
       * @default 0
       */
      indexList: 0,

      /**
       * The speed that the carousel scrolls at in milliseconds.
       * @name gemini.carousel#scrollSpeed
       * @type Integer
       * @default 500
       */
      scrollSpeed: 350,

      /**
       * Whether to map a list of thumbnails to the corresponding pages. Default
       * class expected is ``carousel__thumbs``.
       * @name gemini.carousel#thumbs
       * @type Boolean
       * @default false
       */
      thumbs: false,

      /**
       * Callback function to run when the item changes
       * @name gemini.carousel#onChange
       * @type function
       * @default false
       */
      onChange: false,

      /**
       * Optionally disable touch support
       * @name gemini.carousel#touchEnabled
       * @type boolean
       * @default true
       */
      touchEnabled: true,

      /**
       * Precompiled Handlebar templates to replace default. Expecting 'nav' for
       * default navigation.
       * @name gemini.carousel#templates
       * @type Object
       * @default {}
       */
      templates: {}
    },

    // Event listeners
    events: {
      'click [data-goto]': '_handleClick'
    },

    // Initiate the plugin
    init: function(){
      var P = this;

      //Extend the templates
      P.T = $.extend(T, P.settings.templates);

      // Init variables
      P.currentItem = P.currentPage = 1;

      // Cache jQuery objects
      P.$carousel = !!P.settings.container ?
                    P.$el.find(P.settings.container) :
                    P.$el;
      P.$carouselLists = P.$el.find('.carousel__list');
      P.carouselList = P.$carouselLists[P.settings.indexList];
      P.$carouselList = $(P.carouselList);
      P.$currentPage = P.$el.find('[data-goto="1"]');
      P.$next = P.$el.find('[data-goto="next"],[data-goto="++"]');
      P.$previous = P.$el.find('[data-goto="prev"],[data-goto="previous"],[data-goto="--"]');

      //Change next/prev button active states
      if(!P._isNext()) {
        P.$next.addClass('is-inactive');
      } else {
        P.$next.removeClass('is-inactive');
      }
      if(!P._isPrevious()) {
        P.$previous.addClass('is-inactive');
      } else {
        P.$previous.removeClass('is-inactive');
      }

      // Update
      P._update();

      // Update on resize
      $.respond.bind('resize', function(e, scrn){
        P._update();
        P.gotoPage(P.currentPage, false);
      });

      // Setup thumbnails
      if (P.settings.thumbs) {
        P.$thumbs = P.$el.find('.carousel__thumbs');
        P.$thumbs.find('a').each(function(i){
          $(this).click(function(e){
            e.preventDefault();
            P.gotoPage(i + 1);
          });
        });
      }

      // Touch Support
      if(P.settings.touchEnabled && $.support.touch) {
        P.settings.animate = true;
        P._initTouch();
      }

    },

    /**
     * Update the carousel's cached values and render and templates
     * @name gemini.carousel#_update
     * @private
     * @function
     */
    _update: function(){
      var P = this;

      //Update Cache
      P.pageWidth = P.$carouselList.width();
      P.itemWidth = P.$carouselList.children('li:first-child').width();
      P.itemsPerPage = Math.round(P.pageWidth/P.itemWidth);

      P.itemCount = P.$carouselList.children('li').length;
      P.pageCount = Math.ceil( P.itemCount / P.itemsPerPage );

      // Template the pagination
      if(P.settings.pagination){
        P._paginate();
      }
    },

    /**
     * Render the pagination
     * @name gemini.carousel#_paginate
     * @private
     * @function
     */
    _paginate: function(){
      var P = this;

      //Template
      P.pagination = P.T.nav({
        'pageCount': P.pageCount
      });

      //Render
      if(!P.$pagination){
        //First load
        P.$pagination = $('<span/>').append(P.pagination);
        P.$el.append(P.$pagination);
      } else {
        //Replace exisiting
        P.$pagination.html(P.pagination);
      }

      //Cache
      P.$currentPageCount = P.$pagination.find('.carousel__current-page-count');
    },

    /**
     * Callback when a user clicks the carousel
     * @name gemini.carousel#_handleClick
     * @private
     * @function
     * @param {event#object} e Click event object
     * @param {element} target The targeted element
     */
    _handleClick: function(e, target){
      var P = this;
      var goTo = $(target).data('goto');

      if(!goTo) return;

      e.preventDefault();
      //Next
      if(goTo == "next" || goTo == "++")
        P.next();
      //Previous
      else if(goTo == "prev" || goTo == "previous" || goTo == "--")
        P.previous();
      //Go to page
      else P.gotoPage(goTo);
    },

    /**
     * Check if there is a next page
     * @name gemini.carousel#_isNext
     * @private
     * @function
     * @return {boolean} Weather the next page exists
     */
    _isNext: function(){
      return this.settings.loop || this.currentPage != this.pageCount;
    },

    /**
     * Check if there is a previous page
     * @name gemini.carousel#_isPrevious
     * @private
     * @function
     * @return {boolean} Weather the previous page exists
     */
    _isPrevious: function(){
      return this.settings.loop || this.currentPage != 1;
    },

    /**
     * Go to the next page on the carousel
     * @name gemini.carousel#next
     * @function
     */
    next: function(){
      var P = this;
      if(P._isNext()) P.gotoPage(P.currentPage + 1);
    },

    /**
     * Go to the previous page on the carousel
     * @name gemini.carousel#previous
     * @function
     */
    previous: function(){
      var P = this;
      if(P._isPrevious()) P.gotoPage(P.currentPage - 1);
    },

    /**
     * Go to a specific item in the carousel
     * @name gemini.carousel#_gotoItem
     * @private
     * @function
     */
    _gotoItem: function(item, animate){
      if(animate===undefined) animate = true;

      var P = this;
      if(item > P.itemCount) return;

      //Calculate the x offset in pixels
      var $item = P.$carouselList.children('li:nth-child('+item+')'),
          xOffset = $item.offset().left -
                    P.$carouselList.offset().left +
                    P.carouselList.scrollLeft;

      //Whether there are more items to scroll to
      var isMoreItems = $.rightoffold(
        P.$carouselList.find('li').last(),{
          container:P.$carouselList,
          threshold:-P.itemWidth-20
        }
      );

      //Make sure there's something to scroll to
      if ((item > P.currentItem && !isMoreItems) || xOffset < 0) return;

      //Change the item
      P.currentItem = item;
      if(animate){
        P.$carouselLists.animate({
          scrollLeft:xOffset
        }, P.settings.scrollSpeed);
      }else{
        P.$carouselLists.scrollLeft(xOffset);
      }

      setTimeout(
        _.bind(function(){
          $(this).trigger("scroll");
          if(P.settings.onChange) P.settings.onChange.call(P);
        }, P.$carousel),
      animate ? P.settings.scrollSpeed: 0);
    },

    /**
     * Go to a specific page in the carousel
     * @name gemini.carousel#gotoPage
     * @function
     * @param {Integer} page The desired page
     */
    gotoPage: function(page, animate){
      var P = this;

      if(animate===undefined) animate = P.settings.animate;

      // When there is one item in the carousel list, pageCount = 0,
      // so we need to early exit to avoid an infinite loop
      if ( P.pageCount < 1 ) {
        return;
      }

      if(page > P.pageCount) {
        if (P.settings.loop) P.gotoPage(1);
        return;
      } else if (page < 1) {
        if (P.settings.loop) P.gotoPage(P.pageCount - 1);
        return;
      }

      var item = P.itemsPerPage * (page - 1) + 1;

      P._gotoItem(item, animate);

      //Change active page
      P.currentPage = page;
      P.$currentPage.removeClass('is-active');
      P.$currentPage = P.$el.find('[data-goto="'+P.currentPage+'"]');
      P.$currentPage.addClass('is-active');
      if(!!P.$currentPageCount) P.$currentPageCount.html(page);

      //Change next/prev button active states
      if(!P._isNext()) {
        P.$next.addClass('is-inactive');
      } else {
        P.$next.removeClass('is-inactive');
      }
      if(!P._isPrevious()) {
        P.$previous.addClass('is-inactive');
      } else {
        P.$previous.removeClass('is-inactive');
      }
    },

    /**
     * Initiate the touch support for carousel's
     * @name gemini.carousel#_initTouch
     * @private
     * @function
     */
    _initTouch: function(){
      var P = this;

      require(['gemini.touch'], function(){

        //Add touch events
        P.$carouselList.hammer({
          dragBlockHorizontal: true,
          dragLockToAxis: true,
          dragLockMinDistance: 20,
          hold: false,
          tap: false
        }).on('release dragleft dragright', function(ev){

          switch(ev.type) {
            case 'dragright':
            case 'dragleft':
              // stick to the finger
              var pageOffset = (P.currentPage-1)*P.pageWidth;
              var dragOffset = -ev.gesture.deltaX;

              // slow down at the first and last pane
              if((P.currentPage == 1 && ev.gesture.direction == "right") ||
                (P.currentPage == P.pageCount && ev.gesture.direction == "left")) {
                dragOffset *= 0.4;
              }

              P.$carouselLists.scrollLeft(pageOffset + dragOffset);
              break;

            case 'release':
              // check if their finger is moving fast
              if (ev.gesture.velocityX > 0.05) {
                if (ev.gesture.interimDirection == "left") {
                  P.next();
                } else if (ev.gesture.interimDirection == "right") {
                  P.previous();
                }
              // snap carousel base on positioning of page
              } else {
                // more then 50% moved, navigate
                if (Math.abs(ev.gesture.deltaX) > P.pageWidth/2) {
                  if (ev.gesture.direction == 'right') {
                    P.previous();
                  } else {
                    P.next();
                  }
                } else {
                  P.gotoPage(P.currentPage);
                }
              }
              break;
            }
        });
      });
    }

  });

  // Return the jquery object
  // This way you don't need to require both jquery and the plugin
  return $;

});
